qualtran/bloqs/mcmt/bloq_with_specialized_single_qubit_control.py

```diff
@@ -0,0 +1,149 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import Callable, cast, Iterable, Optional, Sequence, TYPE_CHECKING
+
+import numpy as np
+
+if TYPE_CHECKING:
+    from qualtran import AddControlledT, Bloq, BloqBuilder, CtrlSpec, SoquetT
+    from qualtran._infra.controlled import ControlBit
+
+
+def get_ctrl_system_for_bloq_with_specialized_single_qubit_control(
+    *,
+    ctrl_spec: 'CtrlSpec',
+    current_ctrl_bit: Optional['ControlBit'],
+    bloq_without_ctrl: 'Bloq',
+    get_ctrl_bloq_and_ctrl_reg_name: Callable[['ControlBit'], Optional[tuple['Bloq', str]]],
+) -> tuple['Bloq', 'AddControlledT']:
+    """Build the control system for a bloq with a specialized single-qubit controlled variant.
+
+    Uses the provided specialized implementation when a singly-controlled variant of the bloq is
+    requested. When controlled by multiple qubits, the controls are reduced to a single qubit
+    and the singly-controlled bloq is used.
+
+    The user can provide specializations for the bloq controlled by `1` and (optionally) by `0`.
+    The specialization for control bit `1` must be provided.
+    In case a specialization for a control bit `0` is not provided, the default fallback instead,
+    which wraps the bloq using the `Controlled` metabloq.
+
+    Args:
+        ctrl_spec: The control specification
+        current_ctrl_bit: The control bit of the current bloq, one of `0, 1, None`.
+        bloq_without_ctrl: The variant of this bloq without a control.
+        get_ctrl_bloq_and_ctrl_reg_name: A callable that accepts a control bit (`0` or `1`),
+            and returns the controlled variant of this bloq and the name of the control register.
+            If the callable returns `None`, then the default fallback is used.
+    """
+    from qualtran import CtrlSpec, Soquet
+    from qualtran.bloqs.mcmt import ControlledViaAnd
+
+    def _get_default_fallback():
+        current_bloq: 'Bloq'
+        if current_ctrl_bit is None:
+            current_bloq = bloq_without_ctrl
+        else:
+            ctrl_bloq_and_reg = get_ctrl_bloq_and_ctrl_reg_name(current_ctrl_bit)
+            if ctrl_bloq_and_reg is None:
+                raise ValueError(
+                    f"Expected a controlled bloq (self) matching the current control bit {current_ctrl_bit}, got None"
```

Comment:
what does (self) mean here?

Reply:
I'm actually generally confused about this error message. From the docstring it sounds like `get_ctrl_bloq_and_ctrl_reg_name` is allowed to return None. But not if current_ctrl_bit is set to something?

Reply:
it can only return `None` for `cv=0`. For `cv=1` a specialization must be provided. The type is a bit confusing now after merging the two cases into a single callable, but I don't think there's a cleaner way of doing it (using a single callable)

Reply:
One option could be to make the main function private, and only expose the helpers:
```py
def _get_ctrl_sys(unctrl_bloq, ctrl_getter: Callable[[ControlBit], Optional[Bloq]]): ...

def get(unctrl_bloq, ctrl_getter: Callable[[ControlBit], Bloq]): ...
def get_from_list(unctrl_bloq, ctrl_bloq: Bloq, ctrl_0_bloq: Optional[Bloq]=None): ...
```

Reply:
Ah sorry looks like I misunderstood your question.

`get_ctrl_bloq_and_ctrl_reg_name(current_ctrl_bit)` should return the bloq from which this is called (i.e. `self` of the caller), so it doesn't make sense if it returns `None`.